server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -164,6 +228,190 @@ public IndexerSQLMetadataStorageCoordinator(
+                  existingChunks.size()
+              );
+              return null;
+            } else {
+              SegmentIdentifier max = null;
+
+              if (!existingChunks.isEmpty()) {
+                TimelineObjectHolder<String, DataSegment> existingHolder = Iterables.getOnlyElement(existingChunks);
+                for (PartitionChunk<DataSegment> existing : existingHolder.getObject()) {
+                  if (max == null || max.getShardSpec().getPartitionNum() < existing.getObject()
+                                                                                    .getShardSpec()
+                                                                                    .getPartitionNum()) {
+                    max = SegmentIdentifier.fromDataSegment(existing.getObject());
+                  }
+                }
+              }
+
+              final List<SegmentIdentifier> pendings = getPendingSegmentsForIntervalWithHandle(
+                  handle,
+                  dataSource,
+                  interval
+              );
+
+              for (SegmentIdentifier pending : pendings) {
+                if (max == null ||
+                    pending.getVersion().compareTo(max.getVersion()) > 0 ||
+                    (pending.getVersion().equals(max.getVersion())
+                     && pending.getShardSpec().getPartitionNum() > max.getShardSpec().getPartitionNum())) {
+                  max = pending;
+                }
+              }
+
+              if (max == null) {
+                newIdentifier = new SegmentIdentifier(
+                    dataSource,
+                    interval,
+                    maxVersion,
+                    new NumberedShardSpec(0, 0)
+                );
+              } else if (!max.getInterval().equals(interval) || max.getVersion().compareTo(maxVersion) > 0) {
+                log.warn(
+                    "Cannot allocate new segment for dataSource[%s], interval[%s], maxVersion[%s]: conflicting segment[%s].",
+                    dataSource,
+                    interval,
+                    maxVersion,
+                    max.getIdentifierAsString()
+                );
+                return null;
+              } else if (max.getShardSpec() instanceof LinearShardSpec) {
+                newIdentifier = new SegmentIdentifier(
+                    dataSource,
+                    max.getInterval(),
+                    max.getVersion(),
+                    new LinearShardSpec(max.getShardSpec().getPartitionNum() + 1)
+                );
+              } else if (max.getShardSpec() instanceof NumberedShardSpec) {
+                newIdentifier = new SegmentIdentifier(
+                    dataSource,
+                    max.getInterval(),
+                    max.getVersion(),
+                    new NumberedShardSpec(
+                        max.getShardSpec().getPartitionNum() + 1,
+                        ((NumberedShardSpec) max.getShardSpec()).getPartitions()
```

Comment:
should we add another method `ShardSpec allocateNext()` to `ShardSpec` interface so that above 2 conditionals are avoided and individual shard spec can have the control?